metrics-exporter-prometheus/src/exporter/http_listener.rs

```diff
@@ -0,0 +1,131 @@
+use std::{
+    net::{IpAddr, SocketAddr},
+    sync::Arc,
+};
+
+use http_body_util::Full;
+use hyper::{
+    body::{self, Bytes, Incoming},
+    server::conn::http1::Builder as HyperHttpBuilder,
+    service::service_fn,
+    Request, Response, StatusCode,
+};
+use hyper_util::rt::TokioIo;
+use ipnet::IpNet;
+use tokio::net::{TcpListener, TcpStream};
+use tracing::warn;
+
+use crate::{common::BuildError, ExporterFuture, PrometheusHandle};
+
+struct HttpListeningExporter {
+    inner: Arc<Inner>,
+}
+#[derive(Clone)]
+struct Inner {
+    handle: PrometheusHandle,
+    allowed_addresses: Option<Vec<IpNet>>,
+}
+
+impl HttpListeningExporter {
+    async fn serve(&self, listener: std::net::TcpListener) -> Result<(), hyper::Error> {
+        // TODO - address panic possibility
+        let listener = TcpListener::from_std(listener).unwrap();
```

Comment:
Something needs to be done about the possibility of panic.  There's no good way to generate a hyper::Error.

One options is for this crate to switch to a crate-specific Error for the exporter future result error.  This would of course be a breaking change.

Reply:
I was just trying to preserve the previous semantics of failing with a `BuildError` if the port was unavailable, before the future is even created and awaited in a runtime.  This proposal would be a non-breaking change API-wise, but break consumers relying on a `BuildError` when the port is unavailable. 

But it's entirely your call.

Reply:
That's still the case, as far as I can see?

`PrometheusBuilder::build` is what is calling `new_http_listener`. It was already going to return the error for binding the port because the way you have it creates the underlying `std::net::TcpListener` there already. Converting it to a `tokio::net::TcpListener` doesn't re-bind or anything, and it should only throw errors in very specific conditions like not being able to register the FD to epoll or whatever.